velox/functions/lib/Re2Functions.h

```diff
@@ -30,32 +30,95 @@ namespace facebook::velox::functions {
   /// '%', such as ___%, _%__.
   kAtLeastN,
   /// Pattern with no wildcard characters, such as 'presto', 'foo'.
   kFixed,
+  /// Pattern with single wildcard chars(_) & normal chars, such as
+  /// '_pr_es_to_'.
+  kRelaxedFixed,
   /// Fixed pattern followed by one or more '%', such as 'hello%', 'foo%%%%'.
   kPrefix,
+  /// kRelaxedFixed pattern followed by one or more '%', such as '_pr_es_to_%',
+  /// '_pr_es_to_%%%%'.
+  kRelaxedPrefix,
   /// Fixed pattern preceded by one or more '%', such as '%foo', '%%%hello'.
   kSuffix,
+  /// kRelaxedFixed preceded by one or more '%', such as '%_pr_es_to_',
+  /// '%%%_pr_es_to_'.
+  kRelaxedSuffix,
   /// Patterns matching '%{c0}%', such as '%foo%%', '%%%hello%'.
   kSubstring,
   /// Patterns which do not fit any of the above types, such as 'hello_world',
   /// '_presto%'.
   kGeneric,
 };
 
-struct PatternMetadata {
-  PatternKind patternKind;
-  // Contains the length of the unescaped fixed pattern for patterns of kind
-  // kFixed, kPrefix, kSuffix and kSubstring. Contains the count of wildcard
-  // character '_' for patterns of kind kExactlyN and kAtLeastN. Contains 0
-  // otherwise.
-  size_t length;
-  // Contains the unescaped fixed pattern in patterns of kind kFixed, kPrefix,
-  // kSuffix and kSubstring.
-  std::string fixedPattern = "";
+struct SubPatternMetadata {
+  enum class SubPatternKind {
+    kSingleCharWildcard,
+    kAnyCharsWildcard,
+    kLiteralString
+  };
+
+  SubPatternKind kind;
+  /// Pattern string of the kind.
+  /// For kSingleCharWildcard, it could be '___';
+  /// For kAnyCharsWildcard, it could be '%%';
+  /// For kLiteralString, it could be 'abc'.
+  std::string pattern;
+};
+
+class PatternMetadata {
+ public:
+  static PatternMetadata generic();
+  static PatternMetadata atLeastN(size_t length);
+  static PatternMetadata exactlyN(size_t length);
+  static PatternMetadata fixed(std::vector<SubPatternMetadata> subPatterns);
+  static PatternMetadata relaxedFixed(
+      std::vector<SubPatternMetadata> subPatterns);
+  static PatternMetadata prefix(std::vector<SubPatternMetadata> subPatterns);
+  static PatternMetadata relaxedPrefix(
+      std::vector<SubPatternMetadata> subPatterns);
+  static PatternMetadata suffix(std::vector<SubPatternMetadata> subPatterns);
+  static PatternMetadata relaxedSuffix(
+      std::vector<SubPatternMetadata> subPatterns);
+  static PatternMetadata substring(std::vector<SubPatternMetadata> subPatterns);
+
+  const PatternKind patternKind() const {
+    return patternKind_;
+  }
+
+  const size_t length() const {
+    return length_;
+  }
+
+  const std::vector<SubPatternMetadata>& subPatterns() const {
+    return subPatterns_;
+  }
+
+  const std::string& fixedPattern() const {
+    return fixedPattern_;
+  }
+
+ private:
+  PatternMetadata(
+      PatternKind patternKind,
+      size_t length,
+      std::vector<SubPatternMetadata> subPatterns);
+
+  PatternKind patternKind_;
```

Comment:
const for all member variables

please, add empty lines between member variables to visually separate these

Reply:
We have logic like the following in `Re2Functions.cpp`:

```
PatternMetadata patternMetadata = generic();
try {
    patternMetadata = determinePatternMetadata(...);
} catch(...) {
    // Exception handling.
}
```

It depends on `copy assignement constructor`, so we cann't mark PatternMetadata's fields as const.